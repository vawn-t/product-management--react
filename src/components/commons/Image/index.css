.image {
  height: auto;
  max-width: 100%;
}

/* - Logo */
/* - ---------------------------------------------- */
.image-logo {
  background-repeat: no-repeat;
  width: 6.25rem;
  text-indent: -9999px;
}

/* - Icon */
/* - ---------------------------------------------- */
.image-icon {
  display: block;
  background-repeat: no-repeat;
  width: 2.25rem;
  cursor: pointer;
}

.image-error {
  line-height: 1.5rem;
}
