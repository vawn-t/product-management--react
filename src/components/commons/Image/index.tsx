// Libraries
import React, { memo } from 'react';

// Styles
import './index.css';

// Types
import { ImageVariants } from '@common-types/index';

// Images
import { blankImage } from '@assets/index';

export interface ImageProps {
  alt: string;
  imageUrl: string;
  variant?: ImageVariants;
  onClick?: () => void;
}

const Image: React.FC<ImageProps> = ({
  alt,
  imageUrl = blankImage,
  variant = ImageVariants.Default,
  onClick
}) => (
  <img
    className={`image${
      variant !== ImageVariants.Default ? ` image-${variant} ` : ''
    }`}
    alt={alt}
    src={imageUrl}
    onClick={onClick}
  />
);

export default memo(Image);
