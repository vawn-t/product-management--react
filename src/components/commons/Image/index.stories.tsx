// Libraries
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

// Components
import Image from './index';

// Styles
import '@assets/styles/App.css';
import '@assets/styles/variables.css';
import '@assets/styles/reset.css';
import './index.css';

// Images
import logo from '@assets/images/logo.png';
import iPhoneImage from '@assets/images/iphone-13-pro-max-alpine-green-alpine-green-color.jpg';
import deleteIcon from '@assets/images/icons/trash-fill.svg';

// Types
import { ImageVariants } from '@common-types/image';

export default {
  title: 'Components/Image',
  component: Image
} as ComponentMeta<typeof Image>;

const Template: ComponentStory<typeof Image> = (args) => <Image {...args} />;

export const Default = Template.bind({});
Default.args = {
  imageUrl: iPhoneImage,
  alt: 'iPhone 13 promax image'
};

export const Logo = Template.bind({});
Logo.args = {
  imageUrl: logo,
  alt: 'Logo',
  variant: ImageVariants.Logo
};

export const Icon = Template.bind({});
Icon.args = {
  imageUrl: deleteIcon,
  alt: 'Delete icon',
  variant: ImageVariants.Icon
};
