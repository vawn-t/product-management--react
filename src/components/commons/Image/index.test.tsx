// Libraries
import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { create } from 'react-test-renderer';
import '@testing-library/jest-dom';

// Components
import Image, { ImageProps } from './index';

// Image
import { blankImage } from '@assets/index';

// Types
import { ImageVariants } from '@common-types/index';

describe('Image component', () => {
  let container: HTMLElement;

  const defaultProps: ImageProps = {
    alt: 'image',
    imageUrl: blankImage
  };

  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    container.remove();
    jest.clearAllMocks();
    cleanup();
  });

  test('should render correctly', () => {
    const defaultImage = create(<Image {...defaultProps} />).toJSON();
    expect(defaultImage).toMatchSnapshot();

    const logo = create(
      <Image {...defaultProps} variant={ImageVariants.Logo} />
    ).toJSON();
    expect(logo).toMatchSnapshot();

    const icon = create(
      <Image {...defaultProps} variant={ImageVariants.Icon} />
    ).toJSON();
    expect(icon).toMatchSnapshot();
  });

  test('should be click 1 times', () => {
    const myMock = jest.fn();

    render(<Image {...defaultProps} onClick={myMock} />, { container });

    const image: HTMLImageElement = screen.getByRole('img');

    fireEvent.click(image);

    expect(myMock.mock.calls.length).toEqual(1);
  });
});
