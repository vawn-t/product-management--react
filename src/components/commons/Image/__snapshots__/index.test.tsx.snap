// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component should render correctly 1`] = `
<img
  alt="image"
  className="image"
/>
`;

exports[`Image component should render correctly 2`] = `
<img
  alt="image"
  className="image image-logo "
/>
`;

exports[`Image component should render correctly 3`] = `
<img
  alt="image"
  className="image image-icon "
/>
`;
